api/src/main/java/org/apache/iceberg/catalog/SupportsNamespaces.java

```diff
@@ -0,0 +1,113 @@
+ * without throwing {@link NoSuchNamespaceException} when no namespace is found.
+ */
+public interface SupportsNamespaces {
+  /**
+   * Create a namespace in the catalog.
+   *
+   * @param namespace {@link Namespace}.
+   * @throws AlreadyExistsException If the namespace already exists
+   * @throws UnsupportedOperationException If create is not a supported operation
+   */
+  default void createNamespace(Namespace namespace) {
+    createNamespace(namespace, ImmutableMap.of());
+  }
+
+  /**
+   * Create a namespace in the catalog.
+   *
+   * @param namespace a multi-part namespace
+   * @param metadata a string ImmutableMap of properties for the given namespace
+   * @throws AlreadyExistsException If the namespace already exists
+   * @throws UnsupportedOperationException If create is not a supported operation
+   */
+  void createNamespace(Namespace namespace, ImmutableMap<String, String> metadata);
+
+  /**
+   * List top-level namespaces from the catalog.
+   * <p>
+   * If an object such as a table, view, or function exists, its parent namespaces must also exist
+   * and must be returned by this discovery method. For example, if table a.b.t exists, this method
+   * must return ["a"] in the result array.
+   *
+   * @return an List of namespace {@link Namespace} names
+   */
+  default List<Namespace> listNamespaces() {
+    return listNamespaces(Namespace.empty());
+  }
+
+  /**
+   * List  namespaces from the namespace.
+   * <p>
+   * For example, if table a.b.t exists, use 'SELECT NAMESPACE IN a' this method
+   * must return Namepace.of("a","b") {@link Namespace}.
+   *
+   * @return an List of namespace {@link Namespace} names
+   * @throws NoSuchNamespaceException If the namespace does not exist (optional)
+   */
+  List<Namespace> listNamespaces(Namespace namespace) throws NoSuchNamespaceException;
+
+  /**
+   * Load metadata properties for a namespace.
+   *
+   * @param namespace a Namespace.of(name) {@link Namespace}
+   * @return a string map of properties for the given namespace
+   * @throws NoSuchNamespaceException If the namespace does not exist (optional)
+   */
+  Map<String, String> loadNamespaceMetadata(Namespace namespace) throws NoSuchNamespaceException;
+
+  /**
+   * Drop namespace, while the namespace haven't table or sub namespace will return true.
+   *
+   * @param namespace a Namespace.of(name) {@link Namespace}
+   * @return true while drop success.
+   * @throws NoSuchNamespaceException If the namespace does not exist (optional)
+   */
+  boolean dropNamespace(Namespace namespace) throws NoSuchNamespaceException;
+
+  /**
+   * Create a namespace in the catalog.
+   *
+   * @param namespace a multi-part namespace
+   * @param metadata a string ImmutableMap of properties for the given namespace
+   * @throws UnsupportedOperationException If create is not a supported operation
+   */
+  boolean alterNamespace(Namespace namespace, ImmutableMap<String, String> metadata);
```

Comment:
Looks like you changed the method name from `setNamespaceMetadata` to `alterNamespace`, but that doesn't fix the problem I pointed out. I pointed out that the implementation was preserving properties instead of replacing the old set with the new map.

The reason for that behavior -- and the method name that describes it -- is to make the expected result clear and to make it possible to edit the set of properties. For example, if I call `alterNamespace("db", Map("a" -> "1", "c" -> "3"))` and the namespace properties are `Map("a" -> "x", "b" -> "y")` then what happens to `"b"`? Should it be removed or left unchanged? If it should not be changed, then how does the caller remove a property? The name `setNamespaceMetadata` or a better one, `replaceNamespaceMetadata`, makes it clear that the entire set of properties is being replaced.

There are a couple of alternatives to replacing the entire set of namespace metadata that are easier for the caller. The one I prefer is to mirror Spark's API and expose `alterNamespace(Namespace, NamespaceChange...)` where `NamespaceChange` can be `SetProperty` or `RemoveProperty`. That would look like this, with factory methods for the changes:

```java
catalog.alterNamespace(ns, setProperty("a", "1"), removeProperty("b"));
```